FrontRowXP/owncast
# Replay functionality

## Currently

- Only working with local storage.
- Functionality is hard coded to be enabled, there's no CLI flag yet.
- Only working with entire streams (beginning to end), windowing (clips) is not yet added.
- Playlists are generated on the fly each time. They are not cached or saved. Based on performance and feature preference, these playlists can be persisted. 

## Details

- /api/replays returns a list of all the streams that are replayable.

URLs are generated as so:

-  /replay/_streamId_ for master playlists.
- /replay/_streamId_/_configurationId_ for media playlists.

## Next steps

- Add support for windowed clips.
- Add support for S3 storage for all this functionality.
- Add error handling around if we have records for replayable content, but the files don't actually exist.

## Testing in progress stream

1. Start a live stream with this branch.
2. Visit http://localhost:8080/api/replays
3. Note the "manifest " path.
4. Put the manifest url in your livestreaming video player of choice.
5. Play it. It will be treated as an updating (event) style stream.

## Testing completed streams

1. Stop the live stream.
2. Visit http://localhost:8080/api/replays
3. Notice this stream is no longer in progress.
4. Put the manifest url in your livestreaming video player of choice.
5. Play it. It will be treated as a static (VOD) stream and will not update.

---
Thoughts:

- At launch all streams should be marked as no longer live. If somebody killed Owncast while a stream was live the `endTime` would never get set, leaving it to be infinitely "live".
